File: go-backend-interview-prep/01_go_basics/05_control_flow_if_else/main.go
package main

import "fmt"

func main() {
	// --- Basic if/else ---
	temperature := 22

	if temperature > 30 {
		fmt.Println("hot")
	} else if temperature > 15 {
		fmt.Println("comfortable")
	} else {
		fmt.Println("cold")
	}

	// --- If with init statement ---
	// The variable 'length' is scoped to this if/else block only.
	if length := len("backend"); length > 5 {
		fmt.Println("long word, length:", length)
	} else {
		fmt.Println("short word, length:", length)
	}
	// 'length' does NOT exist here -- compile error if used

	// --- Common pattern: check-and-use ---
	if result := multiply(3, 7); result > 20 {
		fmt.Println("product is large:", result)
	}

	// --- No ternary in Go ---
	// In other languages: status = age >= 18 ? "adult" : "minor"
	// In Go, you must use if/else:
	age := 20
	var status string
	if age >= 18 {
		status = "adult"
	} else {
		status = "minor"
	}
	fmt.Println("status:", status)

	// --- Boolean conditions ---
	loggedIn := true
	isAdmin := false

	if loggedIn && isAdmin {
		fmt.Println("admin dashboard")
	} else if loggedIn {
		fmt.Println("user dashboard")
	} else {
		fmt.Println("please log in")
	}
}

func multiply(a, b int) int {
	return a * b
}

File: go-backend-interview-prep/01_go_basics/05_control_flow_if_else/main.py
# Python equivalent of Control Flow: If/Else (compare with main.go)

def multiply(a: int, b: int) -> int:
    return a * b


def main():
    # --- Basic if/else ---
    temperature = 22

    if temperature > 30:
        print("hot")
    elif temperature > 15:
        print("comfortable")
    else:
        print("cold")

    # --- Python has no if-init statement ---
    # Go:  if length := len("backend"); length > 5 { ... }
    # Python: assign first, then check
    length = len("backend")
    if length > 5:
        print("long word, length:", length)
    else:
        print("short word, length:", length)

    # --- Walrus operator (:=) is the closest equivalent (Python 3.8+) ---
    if (result := multiply(3, 7)) > 20:
        print("product is large:", result)

    # --- No ternary? Python HAS one (Go does not) ---
    age = 20
    status = "adult" if age >= 18 else "minor"
    print("status:", status)

    # --- Boolean conditions ---
    logged_in = True
    is_admin = False

    if logged_in and is_admin:
        print("admin dashboard")
    elif logged_in:
        print("user dashboard")
    else:
        print("please log in")


if __name__ == "__main__":
    main()
